### DISCUSSION TITLE
Iceoryx performance expectations and optimal configurations?

### DISCUSSION THREAD

**Discussion by shuhaowu** [eclipse-iceoryx/iceoryx#1819] [Q&A]
I'm looking at using Iceoryx 2.0.2 to do thread-to-thread communications within the same process in a real-time process (single process mode). My use case is most of the time going to be single-consumer-single-producer, although I don't necessarily want to limit myself to that. According to Iceoryx's documentation, it advertises the ability to pass messages with <1us of latency. I'm wondering how is Iceoryx configured for this to occur?

For reference, I've ran [a benchmark](https://github.com/shuhaowu/lockless-queue-perf-test) where I send a single message from one thread to another for 20 minutes at 200Hz. The threads are running in real-time. This was done on a machine with the RT kernel and CPU pinned to C0, which should eliminate any jitter due to CPU frequency scaling. I also tested the scheduling latency of the computer, which is no more than 70us. The result are as follows. I also included using boost's [spsc_queue](https://www.boost.org/doc/libs/1_56_0/doc/html/boost/lockfree/spsc_queue.html) doing the same thing as a reference.

![image](https://user-images.githubusercontent.com/338100/205806884-e1626e31-2886-42b0-9586-25ca30aca749.png)

So according to this, the mean latency for iceoryx is on the order of 3us, with maximum of about 18us. This is good enough for what I'm using it for, but it does raise a few questions:

- My measurement of 3us is higher than the advertised of ~0.9us. Is the 0.9us result an older result? Maybe it's just my hardware?
- The jitter of up to 15us (18 - 3) seems high. Is this expected? Or perhaps this something to do with my hardware platform?
- The mean latency of iceoryx is an order of magnitude higher than boost spsc queue. I suppose this is expected because boost's queue is limited to SPSC and thus significantly simpler and less costly. Are there any configuration variables that I can tune to improve Iceoryx's performance?

Sorry for the lengthy question and thanks in advanced.

**Comment by elfenpiff** [comment]
@shuhaowu The jittering you observed is known and fixed on iceoryx master, see: https://github.com/eclipse-iceoryx/iceoryx/discussions/1374

The cause was the keep-alive monitoring which send a message via unix domain sockets which caused the jittering. This jittering should not occur as long as you do not start `iox-roudi` with enabled monitoring.

You can measure the latency yourself with the https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/iceperf example. The setup is that leader sends a message to the follower and the follower response to this message. The time from sending the message until the response is arriving is measured and divided by 2 to calculate the latency (2 since we are sending 2 messages).

When I run the benchmark myself I have an average latency of ~1.2us. Could you please hint me where you saw the statement with the 0.9us? I think the number alone is not accurate enough (bad style on our side) and here I would make it more reproducible for the public. I would remeasure everything on a defined hardware like the raspberry pi 4b with raspbian so that everyone can reproduce it.

**Comment by shuhaowu** [comment]
Thanks! I'll take a look at the master branch.

About the 0.9us figure, I'm looking at this graph and reading it approximately:

![](
https://user-images.githubusercontent.com/56729607/157733625-e3e61617-2f72-46ba-b4f2-5a3973f9ad6f.png)
